Mathlib/RingTheory/LaurentSeries.lean

```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+  apply Filter.mem_of_superset hN (fun _ ha ↦ ha _ hn)
+
+/- The support of `Cauchy.coeff` is bounded below -/
+theorem Cauchy.coeff_support_bddBelow {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ) :
+    BddBelow (coeff hℱ).support := by
+  refine ⟨(exists_lb_support hℱ).choose, fun d hd ↦ ?_⟩
+  by_contra hNd
+  exact hd ((exists_lb_support hℱ).choose_spec d (not_le.mp hNd))
+
+/-- To any Cauchy filter ℱ of `laurent_series K`, we can attach a laurent series that is the limit
+of the filter. Its `d`-th coefficient is defined as the limit of `ℱ.coeff d`, which is again Cauchy
+but valued in the discrete space `K`. That sufficiently negative coefficients vanish follows from
+`Cauchy.coeff_support_bddBelow` -/
+def Cauchy.mk_LaurentSeries {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ) : LaurentSeries K :=
+  HahnSeries.mk (coeff hℱ) <| Set.IsWF.isPWO (coeff_support_bddBelow _).wellFoundedOn_lt
+
+/- The following lemma shows that for every `d` smaller than the minimum between the integers
+produced in `cauchy.exists_lb_eventual_support` and `cauchy.exists_lb_support`, for almost all
+series in `ℱ` the `d`th coefficient coincides with the `d`th coefficient of `coeff hℱ`. -/
+theorem Cauchy.exists_lb_coeff_ne {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ) :
+    ∃ N, ∀ᶠ f : LaurentSeries K in ℱ, ∀ d < N, coeff hℱ d = f.coeff d := by
+  obtain ⟨⟨N₁, hN₁⟩, ⟨N₂, hN₂⟩⟩ := exists_lb_eventual_support hℱ, exists_lb_support hℱ
+  refine ⟨min N₁ N₂, ℱ.3 hN₁ fun _ hf d hd ↦ ?_⟩
+  rw [hf d (lt_of_lt_of_le hd (min_le_left _ _)), hN₂ d (lt_of_lt_of_le hd (min_le_right _ _))]
+
+
+/- Given a Cauchy filter in the Laurent Series and a bound `D`, for almost all series in the filter
+the coefficients below `D` coincide with `Caucy.coeff`-/
+theorem Cauchy.coeff_eventually_equal {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ) {D : ℤ} :
+    ∀ᶠ f : LaurentSeries K in ℱ, ∀ d, d < D → coeff hℱ d = f.coeff d := by
+  -- `φ` sends `d` to the set of Laurent Series having `d`th coefficient equal to `ℱ.coeff`.
+  let φ : ℤ → Set (LaurentSeries K) := fun d ↦ {f | coeff hℱ d = f.coeff d}
+  have intersec :
+    (⋂ n ∈ Set.Iio D, φ n) ⊆ {x : LaurentSeries K | ∀ d : ℤ, d < D → coeff hℱ d = x.coeff d} := by
+    intro _ hf
+    simp only [Set.mem_iInter] at hf
+    exact hf
+  let N := min (exists_lb_coeff_ne hℱ).choose D
+  -- The goal becomes to show that the intersection of all `φ n` (for `n ≤ D`) is in `ℱ`.
+  suffices (⋂ n ∈ Set.Iio D, φ n) ∈ ℱ by
+    exact ℱ.3 this intersec
+  -- We first treat the case where `D` is already so small that all series in `ℱ` have trivial
+  -- coefficient below `D`
+  by_cases H : D < (exists_lb_coeff_ne hℱ).choose
+  · apply ℱ.3 (exists_lb_coeff_ne hℱ).choose_spec
+    simp only [Set.mem_Iio, Set.subset_iInter_iff]
+    intro _ hm _ hd
+    exact hd _ (lt_trans hm H)
+  -- We are left with the case when some coefficients below `D` are still non-zero.
+  · rw [← Set.Iio_union_Ico_eq_Iio (le_of_not_gt H), Set.biInter_union]
```

Comment:
I have the impression that the use of `H` is not that necessary (so that the `by_cases H` earlier maybe too complicated than needed).